src/JustEat.StatsD/V2/StatsDPublisherV2.cs

```diff
@@ -0,0 +1,170 @@
+using System;
+using System.Threading;
+using JustEat.StatsD.EndpointLookups;
+
+namespace JustEat.StatsD.V2
+{
+    public sealed class StatsDPublisherV2 : IStatsDPublisher
+    {
+        private const int DefaultSampleRate = 1;
+
+        private static int _bufferSize = 512; // safe max size of udp packet
+
+        [ThreadStatic]
+        private static byte[] _buffer;
+        private static byte[] Buffer() => _buffer ?? (_buffer = new byte[_bufferSize]);
+
+        [ThreadStatic]
+        private static Random _random;
+        private static Random Random() => _random ?? (_random = new Random());
+
+        private readonly StatsDUtf8Formatter _formatter;
+        private readonly IStatsDTransportV2 _transport;
+
+        public StatsDPublisherV2(StatsDConfiguration configuration)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            var endpointSource = EndpointParser.MakeEndPointSource(
+                configuration.Host, configuration.Port, configuration.DnsLookupInterval);
+
+            _transport = new PooledUdpTransportV2(endpointSource);
+            _formatter = new StatsDUtf8Formatter(configuration.Prefix);
+        }
+
+        public StatsDPublisherV2(StatsDConfiguration configuration, IStatsDTransportV2 transport)
```

Comment:
For example, if we ported this into the existing publisher, you could do something like:

```
public StatsDPublisher(StatsDConfiguration configuration)
{
    // Do "v2" stuff
}

public StatsDPublisher(StatsDConfiguration configuration, IStatsDTransport transport)
{
    // Validation goes here...

    if (transport is IStatsDTransportV2 transport2)
    {
        // Do "v2" stuff with transport2
    }
    else
    {
        // Do "v1" stuff
    }
}

public StatsDPublisher(StatsDConfiguration configuration, IStatsDTransportV2 transport)
{
    // Do "v2" stuff
}
```